pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageDeduplication.java

```diff
@@ -0,0 +1,383 @@
+     * returning a future to track the completion of the task
+     */
+    public CompletableFuture<Void> checkStatus() {
+        if (status == Status.Recovering || status == Status.Removing) {
+            // If there's already a transition happening, check later for status
+            pulsar.getExecutor().schedule(this::checkStatus, 1, TimeUnit.MINUTES);
+            return CompletableFuture.completedFuture(null);
+        }
+
+        return isDeduplicationEnabled().thenCompose(shouldBeEnabled -> {
+            if (status != Status.Disabled && !shouldBeEnabled) {
+                // Disabled deduping
+                CompletableFuture<Void> future = new CompletableFuture<>();
+                status = Status.Removing;
+                managedLedger.asyncDeleteCursor(PersistentTopic.DEDUPLICATION_CURSOR_NAME, new DeleteCursorCallback() {
+
+                    @Override
+                    public void deleteCursorComplete(Object ctx) {
+                        status = Status.Disabled;
+                        managedCursor = null;
+                        highestSequencedPushed.clear();
+                        highestSequencedPersisted.clear();
+                        future.complete(null);
+                        log.info("[{}] Disabled deduplication", topic.getName());
+                    }
+
+                    @Override
+                    public void deleteCursorFailed(ManagedLedgerException exception, Object ctx) {
+                        log.warn("[{}] Failed to disable deduplication: {}", topic.getName(), exception.getMessage());
+                        status = Status.Failed;
+                        future.completeExceptionally(exception);
+                    }
+                }, null);
+
+                return future;
+            } else if (status != Status.Enabled && shouldBeEnabled) {
+                // Enable deduping
+                CompletableFuture<Void> future = new CompletableFuture<>();
+                managedLedger.asyncOpenCursor(PersistentTopic.DEDUPLICATION_CURSOR_NAME, new OpenCursorCallback() {
+
+                    @Override
+                    public void openCursorComplete(ManagedCursor cursor, Object ctx) {
+                        // We don't want to retain cache for this cursor
+                        cursor.setInactive();
+                        managedCursor = cursor;
+                        recoverSequenceIdsMap().thenRun(() -> {
+                            status = Status.Enabled;
+                            future.complete(null);
+                            log.info("[{}] Enabled deduplication", topic.getName());
+                        }).exceptionally(ex -> {
+                            status = Status.Failed;
+                            log.warn("[{}] Failed to enable deduplication: {}", topic.getName(), ex.getMessage());
+                            future.completeExceptionally(ex);
```

Comment:
I think `status` should be part of client metrics to verify that dedup is not enabled.

Reply:
You mean, reported in the topic stats ?

Reply:
yes, it would be useful to know in case `status=failed`. so, does it make sense to add in topic-stats?

Reply:
added